aws.go

```diff
@@ -0,0 +1,289 @@
+			break
+		}
+	}
+
+	if exists {
+		hash := md5.New()
+		io.Copy(hash, file)
+		sum := fmt.Sprintf("\"%x\"", hash.Sum(nil))
+
+		head, err := a.client.HeadObject(&s3.HeadObjectInput{
+			Bucket: aws.String(a.vargs.Bucket),
+			Key:    aws.String(remotePath),
+		})
+		if err != nil {
+			return err
+		}
+
+		if sum == *head.ETag {
+			shouldCopy := false
+
+			if head.ContentType == nil && contentType != "" {
+				debug("Content-Type has changed from unset to %s\n", contentType)
+				shouldCopy = true
+			}
+
+			if !shouldCopy && head.ContentType != nil && contentType != *head.ContentType {
+				debug("Content-Type has changed from %s to %s\n", *head.ContentType, contentType)
+				shouldCopy = true
+			}
+
+			if !shouldCopy && len(head.Metadata) != len(metadata) {
+				debug("Count of metadata values has changed for %s\n", localPath)
+				shouldCopy = true
+			}
+
+			if !shouldCopy && len(metadata) > 0 {
+				for k, v := range metadata {
+					if hv, ok := head.Metadata[k]; ok {
+						if *v != *hv {
+							debug("Metadata values have changed for %s\n", localPath)
+							shouldCopy = true
+							break
+						}
+					}
+				}
+			}
+
+			if !shouldCopy {
+				grant, err := a.client.GetObjectAcl(&s3.GetObjectAclInput{
+					Bucket: aws.String(a.vargs.Bucket),
+					Key:    aws.String(remotePath),
+				})
+				if err != nil {
+					return err
+				}
+
+				previousAccess := "private"
+				for _, g := range grant.Grants {
+					gt := *g.Grantee
+					if gt.URI != nil {
+						if *gt.URI == "http://acs.amazonaws.com/groups/global/AllUsers" {
+							if *g.Permission == "READ" {
+								previousAccess = "public-read"
+							} else if *g.Permission == "WRITE" {
+								previousAccess = "public-read-write"
+							}
+						} else if *gt.URI == "http://acs.amazonaws.com/groups/global/AllUsers" {
+							if *g.Permission == "READ" {
+								previousAccess = "authenticated-read"
+							}
+						}
+					}
+				}
+
+				if previousAccess != access {
+					debug("Permissions for \"%s\" have changed from \"%s\" to \"%s\"\n", remotePath, previousAccess, access)
+					shouldCopy = true
+				}
+			}
+
+			if !shouldCopy {
+				debug("Skipping \"%s\" because hashes and metadata match\n", localPath)
+				return nil
+			}
+
+			fmt.Printf("Updating metadata for \"%s\" Content-Type: \"%s\", ACL: \"%s\"\n", localPath, contentType, access)
+			_, err = a.client.CopyObject(&s3.CopyObjectInput{
+				Bucket:            aws.String(a.vargs.Bucket),
+				Key:               aws.String(remotePath),
+				CopySource:        aws.String(fmt.Sprintf("%s/%s", a.vargs.Bucket, remotePath)),
+				ACL:               aws.String(access),
+				ContentType:       aws.String(contentType),
+				Metadata:          metadata,
+				MetadataDirective: aws.String("REPLACE"),
+			})
+			return err
+		}
+
+		_, err = file.Seek(0, 0)
+		if err != nil {
+			return err
+		}
+	}
+
+	fmt.Printf("Uploading \"%s\" with Content-Type \"%s\" and permissions \"%s\"\n", localPath, contentType, access)
```

Comment:
you can use `%q` to quote string instead of `\"%s\"`. The existing code you've written is fine, just figured I'd share the shortcut